### ISSUE TITLE
CastNullableToNonNullableType: false positive on possibly nullable type parameters

### ISSUE THREAD

**Issue by eygraber** [detekt/detekt#6070]
## Expected Behavior
The following code should not violate `CastNullableToNonNullableType`:

```kotlin
inline fun <T, R> combine(
  args: Array<*>,
  crossinline transform: (T) -> R
): R = transform(
  args[0] as T
)
```

## Observed Behavior
It does violate it

## Your Environment
* Version of detekt used: 1.23.0-RC3

**Comment by atulgpt** [comment]
Let me look into it
